analyses/upstream-analysis/util/function-process-Seurat.R

```diff
@@ -0,0 +1,322 @@
+#' @param num_pcs
+#' @param prefix
+#' @param num_dim
+#' @param num_neighbors
+#' @param results_dir
+#' @param plots_output
+#' @param use_condition_split
+#' @param condition
+#' @param print_pdf
+#' @param PCA_Feature_List
+#'
+#' @return
+#' @export
+#'
+#' @examples
+#' 
+Process_Seurat <- function(seurat_obj, nfeatures_value, Genome, Regress_Cell_Cycle, assay, 
+                          num_pcs, prefix, num_dim, num_neighbors, results_dir, 
+                          plots_output, use_condition_split, condition, print_pdf, PCA_Feature_List) {
+  
+  set.seed(1234) # Make code reproducible
+
+  seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 10000, verbose = TRUE) %>% 
+    FindVariableFeatures(selection.method = "vst", nfeatures = nfeatures_value) 
+  
+  cat("Get cell cycle scores and then scale data", "\n")
+  # Get cell cycle scores (S and G2M) for each cell, Genome specifies which gene set to use for cell cycle genes
+  seurat_obj <- Genome_Specific_Cell_Cycle_Scoring(seurat_obj, Genome)
+  
+  cat("Scale data", "\n")
+  # Indicates whether or not to regress for cell cycle and, if so, which method to use and scale data
+  if (Regress_Cell_Cycle == "NO") {
+    print("We will NOT regress out for cell cycle and scale data!")
+    seurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))
+    } else if (Regress_Cell_Cycle == "YES") {    
+      print("We will regress out for cell cycle and scale data!")
+      seurat_obj <- ScaleData(seurat_obj, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(seurat_obj))
+      } else if (Regress_Cell_Cycle == "DIFF") {
+        print("We will regress out for cell cycle difference and scale data!")
+        seurat_obj$CC.Difference <- seurat_obj$S.Score - seurat_obj$G2M.Score
+        seurat_obj <- ScaleData(seurat_obj, vars.to.regress = "CC.Difference", features = rownames(seurat_obj))
+  } 
+  
+  cat("Run PCA", "\n") # Get PCs for dimensionality reduction
+  # Either use the prespecified PCA.Feature.List or use Variable Features found before
+  if (missing(PCA_Feature_List)) {
+    seurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj), assay = assay, npcs = num_pcs, verbose = FALSE)
+  } else {
+      seurat_obj <- RunPCA(seurat_obj, features = PCA_Feature_List, assay = assay, npcs = num_pcs, verbose = FALSE)
+  }
+  
+  cat("Run UMAP", "\n")
+  set.seed(1234) # Make code reproducible
+  
+  # Create a dataframe of all of the possible combinations of number of PCs to use for UMAP, and the number of neighbors
+  num_dim_vect <- c(num_dim)
+  num_neighbors_vect <- c(num_neighbors)
+  possibilities <- expand.grid(num_dim_vect, num_neighbors_vect)
+  
+  # For each of these combinations, let's calculate UMAP
+  for(k in 1:nrow(possibilities)) {
+    num_dim <- possibilities[k, 1]
+    num_neighbors <- possibilities[k, 2]
+    seurat_obj <- run_dr(data = seurat_obj, dr_method = "umap", reduction = paste0("pca"),
```

Comment:
I assume the run_dr script is defined in another script and/or module? And that is where the code for things like finding neighbors, generating clusters, and doing UMAP reduction is found? Just double checking for my own clarity.

Reply:
The `run_dir` function is called from the `scooter` R package loaded in the relative script.

Reply:
@AntoniaChroni Maybe for extra clarity, this command could be scooter::run_dr, so people looking at the code can understand where the function is from if it's not a function you define. I myself was not familiar with the scooter package. Also, again for my own understanding, what is the purpose for only doing dimensionality reductions and not having a step for finding clusters?

Reply:
@sharonfreshour During the `upstream-analysis`, we are interested in exploring the quality of the library, filtering low-quality cells via the best methods fitting the data, and data visualization. The `Dimensionality reduction` step allows the user to do the latter. We use various numbers of PCs and numbers of neighbors for UMAP by utilizing the `run_dr` function as we want to explore the dimensionality per library at the first step of filtering.

Cluster determination is used for biological purposes and questions, e.g., cell type identification, once the data are **fully filtered** (which is not the case here). This will be done in a separate module once we have our data fully filtered and ready for `downstream-analysis`.

Reply:
I see, so these functions will not be reused in downstream processing, that makes sense.